kernel/src/table_properties/deserialize.rs

```diff
@@ -0,0 +1,287 @@
+        byte_buf option unit unit_struct newtype_struct seq tuple tuple_struct
+        map struct enum identifier ignored_any
+    }
+}
+
+struct HashMapMapAccess<'de> {
+    iter: std::collections::hash_map::Iter<'de, String, String>,
+    value: Option<&'de String>,
+}
+
+impl<'de> MapAccess<'de> for HashMapMapAccess<'de> {
+    type Error = de::value::Error;
+
+    fn next_key_seed<K>(&mut self, seed: K) -> Result<Option<K::Value>, Self::Error>
+    where
+        K: DeserializeSeed<'de>,
+    {
+        match self.iter.next() {
+            Some((key, value)) => {
+                self.value = Some(value);
+                let de = de::value::StrDeserializer::new(key);
+                seed.deserialize(de).map(Some)
+            }
+            None => Ok(None),
+        }
+    }
+
+    fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value, Self::Error>
+    where
+        V: DeserializeSeed<'de>,
+    {
+        let value = self.value.take().unwrap(); // FIXME
+        let de = de::value::StrDeserializer::new(value);
+        seed.deserialize(de)
+    }
+}
+
+pub(crate) fn deserialize_option<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
+where
+    D: Deserializer<'de>,
+    T: Deserialize<'de>,
+{
+    T::deserialize(deserializer).map(Some)
+}
+
+pub(crate) fn deserialize_pos_int<'de, D>(deserializer: D) -> Result<Option<u64>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    let n: u64 = s.parse().unwrap(); // FIXME
+    if n == 0 {
+        panic!("FIXME");
+        // return Err("something");
+    }
+    Ok(Some(n))
+}
+
+pub(crate) fn deserialize_bool<'de, D>(deserializer: D) -> Result<Option<bool>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    match s.as_str() {
+        "true" => Ok(Some(true)),
+        "false" => Ok(Some(false)),
+        _ => Err(de::Error::unknown_variant(&s, &["true", "false"])),
+    }
+}
+
+pub(crate) fn deserialize_column_names<'de, D>(
+    deserializer: D,
+) -> Result<Option<Vec<ColumnName>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    Ok(Some(
+        s.split(',')
+            .map(|name: &str| ColumnName::new([name]))
+            .collect(),
+    ))
+}
+
+pub(crate) fn deserialize_interval<'de, D>(deserializer: D) -> Result<Option<Duration>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    parse_interval(&s).map(Some).map_err(de::Error::custom)
+}
+
+fn parse_interval(value: &str) -> Result<Duration, String> {
+    let not_an_interval = || format!("'{value}' is not an interval");
+
+    if !value.starts_with("interval ") {
+        return Err(not_an_interval());
+    }
+    let mut it = value.split_whitespace();
+    let _ = it.next(); // skip "interval"
+    let number = parse_int(it.next().ok_or_else(not_an_interval)?)?;
```

Comment:
is there a particular thing we're trying to catch by parsing as a signed number and then returning an error if it's negative? would it not make more sense to just parse as a `u64` and fail if we fail to parse?

I guess the error message is a little nicer in this case. Not sure it's worth the extra code.

Reply:
happy to delete if not worth it